File: core/src/com/bombbird/terminalcontrol/sounds/Pronunciation.java
package com.bombbird.terminalcontrol.sounds;

import com.bombbird.terminalcontrol.utilities.saving.FileLoader;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;

import java.util.HashMap;

public class Pronunciation {
    public static final HashMap<String, String> waypointPronunciations = new HashMap<String, String>();
    public static HashMap<String, String> callsigns;

    /** Loads the pronunciations */
    public static void loadPronunciation() {
        waypointPronunciations.put("APU", "aun poo");
        waypointPronunciations.put("YILAN", "yee lan");
        waypointPronunciations.put("HLG", "ho long");
        waypointPronunciations.put("ZUH", "Lian Sheng Way");
        callsigns = FileLoader.loadIcaoCallsigns();
    }

    /** Checks whether direct contains a number */
    public static String checkNumber(String direct) {
        String[] numbers = {"1", "2", "3", "4", "5", "6", "7", "8", "9", "0"};
        String[] directList = direct.split("");
        for (int i = directList.length - 1; i>= 0; i--) {
            if (ArrayUtils.contains(numbers, directList[i])) return StringUtils.join(directList, " ");
        }
        return direct;
    }

    /** Converts any number into text due to special pronunciation requirements for 0, 9 and . */
    public static String convertNoToText(String altitude) {
        String[] list = altitude.split("");
        for (int i = 0; i < list.length; i++) {
            if (list[i].equals("0")) list[i] = "zero";
            if (list[i].equals("9")) list[i] = "niner";
            if (list[i].equals(".")) list[i] = "decimal";
        }
        return StringUtils.join(list, " ");
    }

    /** Converts any FLXXX in text to flight level X X X */
    public static String convertToFlightLevel(String action) {
        String[] actionList = action.split(" ");
        for (int i = 0; i < actionList.length; i++) {
            if (actionList[i].contains("FL")) {
                String altitude = Pronunciation.convertNoToText(actionList[i].substring(2));
                String phrase = "flight level" + altitude;
                actionList[i] = phrase;
            }
        }
        return StringUtils.join(actionList, " ");
    }
}

File: core/src/com/bombbird/terminalcontrol/sounds/Pronunciation.kt
package com.bombbird.terminalcontrol.sounds

import com.bombbird.terminalcontrol.utilities.saving.FileLoader
import org.apache.commons.lang3.StringUtils
import kotlin.collections.HashMap

object Pronunciation {
    val waypointPronunciations = HashMap<String, String>()
    val alphabetPronunciations = HashMap<Char, String>()
    val numberPronunciations = HashMap<Char, String>()
    lateinit var callsigns: HashMap<String, String>

    /** Loads the pronunciations  */
    fun loadPronunciation() {
        waypointPronunciations["PAU"] = "pan on"
        waypointPronunciations["LHG"] = "low hon"
        waypointPronunciations["UZH"] = "sheng lian"
        waypointPronunciations["LNG"] = "lung nan"
        waypointPronunciations["MST"] = "mo she toe"
        waypointPronunciations["TVB"] = "tilla vobas"
        waypointPronunciations["LTD"] = "tornado"
        waypointPronunciations["VNS"] = "vanas"
        waypointPronunciations["JCN"] = "jo castin"
        waypointPronunciations["BRO"] = "noble ditto"
        waypointPronunciations["ISE"] = "momo terra"
        alphabetPronunciations['A'] = "alpha"
        alphabetPronunciations['B'] = "bravo"
        alphabetPronunciations['C'] = "charlie"
        alphabetPronunciations['D'] = "delta"
        alphabetPronunciations['E'] = "echo"
        alphabetPronunciations['F'] = "foxtrot"
        alphabetPronunciations['G'] = "golf"
        alphabetPronunciations['H'] = "hotel"
        alphabetPronunciations['I'] = "india"
        alphabetPronunciations['J'] = "juliett"
        alphabetPronunciations['K'] = "kilo"
        alphabetPronunciations['L'] = "lima"
        alphabetPronunciations['M'] = "mike"
        alphabetPronunciations['N'] = "november"
        alphabetPronunciations['O'] = "oscar"
        alphabetPronunciations['P'] = "papa"
        alphabetPronunciations['Q'] = "quebec"
        alphabetPronunciations['R'] = "romeo"
        alphabetPronunciations['S'] = "sierra"
        alphabetPronunciations['T'] = "tango"
        alphabetPronunciations['U'] = "uniform"
        alphabetPronunciations['V'] = "victor"
        alphabetPronunciations['W'] = "whiskey"
        alphabetPronunciations['X'] = "x-ray"
        alphabetPronunciations['Y'] = "yankee"
        alphabetPronunciations['Z'] = "zulu"
        numberPronunciations['0'] = "zero"
        numberPronunciations['1'] = "one"
        numberPronunciations['2'] = "two"
        numberPronunciations['3'] = "three"
        numberPronunciations['4'] = "four"
        numberPronunciations['5'] = "five"
        numberPronunciations['6'] = "six"
        numberPronunciations['7'] = "seven"
        numberPronunciations['8'] = "eight"
        numberPronunciations['9'] = "niner"
        numberPronunciations['.'] = "decimal"
        callsigns = FileLoader.loadIcaoCallsigns()
    }

    /** Checks whether direct contains a number  */
    fun checkNumber(direct: String): String {
        val directList = direct.toCharArray()
        for (i in directList) {
            if (numberPronunciations.containsKey(i)) return StringUtils.join(directList, " ")
        }
        return direct
    }

    /** Converts any number into text due to special pronunciation requirements for 0, 9 and .  */
    fun convertNoToText(number: String): String {
        val list = number.toCharArray()
        val newList = ArrayList<String>()
        for (i in list) {
            newList.add(numberPronunciations[i] ?: i.toString())
        }
        return StringUtils.join(newList, " ")
    }

    /** Converts any FLXXX in text to flight level X X X, returns joined text  */
    fun convertToFlightLevel(action: String): String {
        val actionList = action.split(" ".toRegex()).toTypedArray()
        for (i in actionList.indices) {
            if (actionList[i].contains("FL")) {
                val altitude = convertNoToText(actionList[i].substring(2))
                val phrase = "flight level $altitude"
                actionList[i] = phrase
            }
        }
        return StringUtils.join(actionList, " ")
    }
}